src/librustc_typeck/astconv.rs

```diff
@@ -1133,14 +1123,32 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx> + 'o {
             })
         });
 
-        // Dedup auto traits so that `dyn Trait + Send + Send` is the same as `dyn Trait + Send`.
-        auto_traits.sort();
-        auto_traits.dedup();
+        // HACK(#57057): see below.
+        let extra_principal_trait = auto_traits.iter().skip(1).any(
+            |tr| tr.trait_ref().def_id() == principal.def_id());
+
+        // Remove duplicate auto traits.
```

Comment:
If we removed the auto traits earlier, then we wouldn't need this step?

Reply:
No this is needed for compatibility with the previous bug in the compiler.

Reply:
No this is needed for compatibility with the previous bug in the compiler.

Reply:
Actually, I misunderstood. This is all necessary because of collecting the entire set of auto traits. If we ignored auto traits in the above parts of code, auto traits would get ignored entirely in trait objects, which is obviously very bad. I'll remove the comment about linting, since that's misleading.

Reply:
Yeah. It's needed for both the bug *and* linting, I think. Though obviously we're not doing linting now, it's worth leaving that comment in, for whoever implements linting.

Reply:
Honestly, I don't understand how this code works exactly, but I highly recommend not changing it, because it took me hours to figure out, and this is all very subtle business. I think we should leave it as is. The FIXME comment could be misleading. But you can make the changes yourself if you think you know.